illumos-utils/src/running_zone.rs

```diff
@@ -177,13 +185,49 @@ mod zenter {
         fn ct_tmpl_activate(fd: c_int) -> c_int;
         fn ct_tmpl_clear(fd: c_int) -> c_int;
         fn ct_ctl_abandon(fd: c_int) -> c_int;
+        fn ct_event_read_critical(fd: c_int, ev: *mut ct_evthdl_t) -> c_int;
+        fn ct_event_get_type(ev: ct_evthdl_t) -> u64;
+        fn ct_event_get_ctid(ev: ct_evthdl_t) -> ctid_t;
+        fn ct_event_free(ev: ct_evthdl_t);
     }
 
     #[link(name = "c")]
     extern "C" {
         pub fn zone_enter(zid: zoneid_t) -> c_int;
     }
 
+    // This thread watches for critical events coming from all process
+    // contracts held by sled-agent, and reaps (abandons) contracts which
+    // become empty. Process contracts are used in conjunction with
+    // zone_enter() in order to run commands within non-global zones, and
+    // the contracts used for this come from templates that define becoming
+    // empty as a critical event.
+    pub fn contract_reaper(log: &Logger) {
+        const EVENT_PATH: &[u8] = b"/system/contract/process/pbundle";
+        const CT_PR_EV_EMPTY: u64 = 1;
+
+        let cpath = CString::new(EVENT_PATH).unwrap();
+        let fd = unsafe { libc::open(cpath.as_ptr(), libc::O_RDONLY) };
+
+        loop {
+            let mut ev: ct_evthdl_t = std::ptr::null_mut();
+            let evp: *mut ct_evthdl_t = &mut ev;
+            if unsafe { ct_event_read_critical(fd, evp) } == 0 {
```

Comment:
Are there any error returns from this that indicate it will never succeed? If we hit those, will we go into a tight loop of forever failing?

Reply:
Not according to the man page, but the library can return transient errors such as ENOMEM. I've updated this.